File: Dijkstra.hs
import Data.List (minimumBy)
import Data.Ord (comparing)

type Graph = [[Int]]
type Distance = Int

inf :: Int
inf = maxBound `div` 2

dijkstra :: Graph -> Int -> [Distance]
dijkstra graph start = dijkstra' initialDist initialVisited
  where
    n = length graph
    initialDist = [if i == start then 0 else inf | i <- [0..n-1]]
    initialVisited = replicate n False

    dijkstra' dist visited
      | all id visited = dist
      | otherwise =
          let u = minimumBy (comparing (dist !!))
                  [i | i <- [0..n-1], not (visited !! i)]
              visited' = update visited u True
              dist' = foldl (relax u) dist [0..n-1]
          in dijkstra' dist' visited'

    relax u d v
      | visited !! v = d
      | graph !! u !! v == 0 = d
      | d !! u + graph !! u !! v < d !! v =
          update d v (d !! u + graph !! u !! v)
      | otherwise = d

    update xs i val =
      take i xs ++ [val] ++ drop (i + 1) xs

main :: IO ()
main = do
  let graph =
        [ [0, 4, 0, 0, 6]
        , [4, 0, 2, 0, 0]
        , [0, 2, 0, 5, 0]
        , [0, 0, 5, 0, 3]
        , [6, 0, 0, 3, 0]
        ]

  let start = 0
  let distances = dijkstra graph start

  putStrLn ("Distances depuis le sommet " ++ show start)
  mapM_ (\(i,d) -> putStrLn ("Vers " ++ show i ++ " : " ++ show d))
        (zip [0..] distances)

File: dijkstra.ex
defmodule Dijkstra do
  @inf 1_000_000_000

  def dijkstra(graph, start) do
    n = length(graph)
    dist = List.duplicate(@inf, n) |> List.replace_at(start, 0)
    visited = List.duplicate(false, n)

    iterate(graph, dist, visited, n - 1)
  end

  defp iterate(_graph, dist, _visited, 0), do: dist

  defp iterate(graph, dist, visited, k) do
    {u, _} =
      dist
      |> Enum.with_index()
      |> Enum.filter(fn {_, i} -> not Enum.at(visited, i) end)
      |> Enum.min_by(fn {d, _} -> d end, fn -> {@inf, -1} end)

    if u == -1 do
      dist
    else
      visited = List.replace_at(visited, u, true)
      dist = relax(graph, dist, visited, u)
      iterate(graph, dist, visited, k - 1)
    end
  end

  defp relax(graph, dist, visited, u) do
    Enum.with_index(dist)
    |> Enum.map(fn {d, v} ->
      w = graph |> Enum.at(u) |> Enum.at(v)

      cond do
        Enum.at(visited, v) -> d
        w == 0 -> d
        Enum.at(dist, u) + w < d -> Enum.at(dist, u) + w
        true -> d
      end
    end)
  end
end

# Exemple d'utilisation
graph = [
  [0, 4, 0, 0, 1],
  [4, 0, 2, 0, 0],
  [0, 2, 0, 3, 0],
  [0, 0, 3, 0, 5],
  [1, 0, 0, 5, 0]
]

start = 0
distances = Dijkstra.dijkstra(graph, start)

IO.puts("Distances depuis le sommet #{start}")
Enum.with_index(distances)
|> Enum.each(fn {d, i} ->
  IO.puts("Vers #{i} : #{d}")
end)
